pkg/envoy/resources.go

```diff
@@ -27,9 +34,63 @@ const (
 	NetworkPolicyHostsTypeURL = "type.googleapis.com/cilium.NetworkPolicyHosts"
 )
 
+// NPHDSCache is a cache of resources in the Network Policy Hosts Discovery
+// Service.
+type NPHDSCache struct {
+	*xds.Cache
+}
+
+func newNPHDSCache() NPHDSCache {
+	return NPHDSCache{Cache: xds.NewCache()}
+}
+
 var (
 	// NetworkPolicyHostsCache is the global cache of resources of type
 	// NetworkPolicyHosts. Resources in this cache must have the
 	// NetworkPolicyHostsTypeURL type URL.
-	NetworkPolicyHostsCache = xds.NewCache()
+	NetworkPolicyHostsCache = newNPHDSCache()
 )
+
+// OnIPIdentityCacheGC is required to implement IPIdentityMappingListener.
+func (cache *NPHDSCache) OnIPIdentityCacheGC() {
+	// We don't have anything to synchronize in this case.
+}
+
+// OnIPIdentityCacheChange pushes modifications to the IP<->Identity mapping
+// into the Network Policy Host Discovery Service (NPHDS).
+func (cache *NPHDSCache) OnIPIdentityCacheChange(
+	modType ipcache.CacheModification, ipIDPair identity.IPIdentityPair) {
+
+	if modType == ipcache.Upsert {
+		endpointIPs, _ := ipcache.IPIdentityCache.LookupByIdentity(ipIDPair.ID)
+
+		// Update XDS Cache as well with the new identity to IP mapping.
+		// The old identity to IP mapping for this IP in the XDS
+		// cache may still be present.
+		ipStrings := make([]string, 0, len(endpointIPs))
+		for endpointIP := range endpointIPs {
+			ipStrings = append(ipStrings, endpointIP)
+		}
+		sort.Strings(ipStrings)
+		npHost := &envoyAPI.NetworkPolicyHosts{Policy: uint64(ipIDPair.ID), HostAddresses: ipStrings}
+		cache.Upsert(NetworkPolicyHostsTypeURL, ipIDPair.ID.StringID(), npHost, false)
+	}
+	cache.updateForOldIdentity(ipIDPair.ID)
+}
+
+// updateXDSCacheForOldIdentity updates the XDS cache entry for old identity
+// with the list of IPs that is stored in the IPIdentityCache.
+func (cache *NPHDSCache) updateForOldIdentity(oldIdentity identity.NumericIdentity) {
+	endpointIPs, isIdentityInCache := ipcache.IPIdentityCache.LookupByIdentity(oldIdentity)
```

Comment:
You shouldn't need to call `ipcache`. You have enough state in `xds.Cache` to determine that locally.

Reply:
Well, not exactly. This code assumes that `ipcache.IPIdentityCache` has been updated with the new IP->ID mapping, but the `NPHDSCache` has not yet been updated (because this code is supposed to update it). So it goes back up to the ipcache to get the latest list of IPs for the identity and pushes this into NPHDS.

Reply:
That's not needed. Again, you have all the information you need here.
You'll probably need to refactor this function and `OnIPIdentityCacheChange` above.
If `modType` is `Delete`, you need to first get the list of IPs associated with the ID from the cache, remove the IP from that lsit, and then: 1) if the new set is empty, delete the entry from the cache, otherwise 2) update the entry in the cache with the new list.

Reply:
I see the point, the interface is broken if the callback from that layer has to go back up to that layer to fetch the relevant information.